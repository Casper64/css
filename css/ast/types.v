module ast

import css.token

pub type RuleType = AtRule | Empty | KeyframesRule | Rule

pub type Prelude = Raw | SelectorList

pub enum AtType {
	unkown
	charset
	container
	counter_style
	font_face
	@import
	keyframes
	layer
	media
}

pub struct NodeError {
pub:
	pos  token.Pos
	msg  string
	code int
}

pub fn (n NodeError) msg() string {
	return n.msg
}

pub fn (n NodeError) code() int {
	return n.code
}

pub type PseudoSelector = PseudoClassSelector | PseudoElementSelector | Raw

pub enum AttributeMatchType {
	@none
	exact // =
	contains // *=
	starts_with // ^=
	ends_with // $=
}

pub enum OperatorKind {
	plus // +
	min // -
	mul // *
	div // /
	comma // ,
}

pub enum ImportType {
	url
	path
}
