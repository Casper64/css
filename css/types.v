module css

import css.ast
import css.datatypes

pub type ColorValue = datatypes.Color | string

// value between 0.0 and 1.0
pub type AlphaValue = Keyword | f64
pub type DimensionValue = Keyword
	| datatypes.CalcSum
	| datatypes.Length
	| datatypes.Percentage
	| f64

pub type Keyword = string

pub type Image = Gradient | Keyword | Url

pub type Value = AlphaValue
	| Background
	| Border
	| BorderColors
	| BorderStyles
	| ColorValue
	| DimensionValue
	| Gradient
	| Image
	| Keyword
	| MarginPadding
	| Overflow
	| Shadow
	| Text
	| TextCombineUpright
	| TextEllipsis
	| TextOverflow
	| string

pub struct Attribute {
pub mut:
	name    string
	matcher ast.AttributeMatchType
	value   ?string
}

pub fn (a Attribute) str() string {
	mut s := '[${a.name}'
	if v := a.value {
		s += match a.matcher {
			.@none { '' }
			.exact { '="${v}"' }
			.contains { '*="${v}"' }
			.starts_with { '^="${v}"' }
			.ends_with { '$="${v}"' }
		}
	}
	s += ']'
	return s
}

pub fn (a Attribute) == (b Attribute) bool {
	if a.name != b.name {
		return false
	}

	if va := a.value {
		vb := b.value or { return false }

		return match a.matcher {
			.@none {
				// should never reach this
				false
			}
			.exact {
				va == vb
			}
			.contains {
				va.contains(vb)
			}
			.starts_with {
				va.starts_with(vb)
			}
			.ends_with {
				va.ends_with(vb)
			}
		}
	}
	// a[href]
	return true
}

pub type Class = string

pub fn (c Class) str() string {
	return '.' + c
}

pub type Combinator = string

pub fn (cm Combinator) str() string {
	if cm != ' ' {
		return ' ' + cm + ' '
	} else {
		return ' '
	}
}

pub type Id = string

pub fn (id Id) str() string {
	return '#' + id
}

pub type Type = string

pub fn (typ Type) str() string {
	return typ
}

pub struct PseudoClass {
pub mut:
	name     string
	children []Selector
}

pub fn (pc PseudoClass) str() string {
	mut s := ':${pc.name}'
	if pc.children.len > 0 {
		s += '(${pc.children})'
	}
	return s
}

pub struct PseudoElement {
pub mut:
	name     string
	children []Selector
}

pub fn (pe PseudoElement) str() string {
	mut s := '::${pe.name}'
	if pe.children.len > 0 {
		s += '(${pe.children})'
	}
	return s
}

pub type Selector = Attribute | Class | Combinator | Id | PseudoClass | PseudoElement | Type

pub fn (selectors []Selector) str() string {
	mut res := ''
	for s in selectors {
		res += match s {
			Class { s.str() }
			Id { s.str() }
			Type { s.str() }
			Combinator { s.str() }
			PseudoClass { s.str() }
			PseudoElement { s.str() }
			Attribute { s.str() }
		}
	}
	return res
}

// check if `other_selectors` is a subset of `current_selectors`
// matches the part of `current_selectors` after the last Combinator
pub fn (current_selectors []Selector) matches(other_selectors []Selector) bool {
	// last_selectors are all selectors that come after the last Combinator
	// in `current_selectors`
	mut last_selectors := []Selector{}
	for s in current_selectors {
		if s is Combinator {
			last_selectors.clear()
		} else {
			last_selectors << s
		}
	}

	for s in last_selectors {
		match s {
			PseudoClass {
				if s.name == 'not' {
					if s.children.matches(other_selectors) {
						return false
					}
				}
				// TODO: :is
			}
			PseudoElement {
				// pseudo elements should be matched in combination with a DOM
				return true
			}
			Attribute {
				for a in other_selectors {
					if a is Attribute && s != a {
						return false
					}
				}
			}
			else {
				if s !in other_selectors {
					return false
				}
			}
		}
	}
	return true
}

// Specificity represents CSS specificity
// https://developer.mozilla.org/en-US/docs/Web/CSS/Specificity
pub struct Specificity {
pub mut:
	col1 int
	col2 int
	col3 int
}

pub fn (s &Specificity) str() string {
	return '${s.col1}-${s.col2}-${s.col3}'
}

pub fn Specificity.from_selectors(selectors []Selector) Specificity {
	mut s := Specificity{}

	for current_selector in selectors {
		match current_selector {
			Id {
				s.col1++
			}
			Attribute, Class {
				s.col2++
			}
			PseudoClass {
				if current_selector.name == 'where' {
					// :where doesn't count for specificity, same applies for its children
					continue
				} else if current_selector.name !in ['is', 'has', 'not'] {
					// :is, :has and :not don't count for specificity, but their children do
					s.col2++
				}
				s += Specificity.from_selectors(current_selector.children)
			}
			Type {
				if current_selector != '*' {
					// the universal selector does not count for specificity
					s.col3++
				}
			}
			PseudoElement {
				s.col3++
				s += Specificity.from_selectors(current_selector.children)
			}
			else {}
		}
	}

	return s
}

fn (a Specificity) == (b Specificity) bool {
	return a.col1 == b.col1 && a.col2 == b.col2 && a.col3 == b.col3
}

fn (a Specificity) < (b Specificity) bool {
	if a.col1 < b.col1 {
		return true
	} else if a.col1 == b.col1 {
		if a.col2 < b.col2 {
			return true
		} else if a.col2 == b.col2 {
			return a.col3 < b.col3
		}
	}
	return false
}

fn (a Specificity) + (b Specificity) Specificity {
	return Specificity{
		col1: a.col1 + b.col1
		col2: a.col2 + b.col2
		col3: a.col3 + b.col3
	}
}

fn (a Specificity) - (b Specificity) Specificity {
	return Specificity{
		col1: a.col1 - b.col1
		col2: a.col2 - b.col2
		col3: a.col3 - b.col3
	}
}

pub struct RawValue {
pub mut:
	value     Value
	important bool
}

// Rule represents a css rule with only a single selector
pub struct Rule {
pub mut:
	specificity  Specificity
	selectors    []Selector
	declarations map[string]RawValue
}

@[inline]
pub fn (r Rule) matches(selectors []Selector) bool {
	return r.selectors.matches(selectors)
}

pub fn (rules []Rule) get_styles() map[string]Value {
	// TODO: merge grouped properties together e.g. `background` gets split
	// into 'background-color', 'background-width' etc.
	// OR combine them from `background-color` to `background` only
	mut styles := map[string]Value{}
	mut importants := map[string]bool{}

	// TODO: reduce iterations
	for rule in rules {
		for property, value in rule.declarations {
			if !value.important && importants[property] == false {
				set_grouped(property, value.value, mut styles)
			} else if value.important {
				set_grouped(property, value.value, mut styles)
				importants[property] = true
			}
		}
	}

	return styles
}

pub type TextCombineUprightDigits = int
pub type TextCombineUpright = Keyword | TextCombineUprightDigits

// text-overflow: ellipsis "[..]"; will become `css.TextOverflow(css.TextEllipsis('[..]'))`
pub type TextEllipsis = string
pub type TextOverflow = Keyword | TextEllipsis

pub type ShadowValue = Keyword | Shadow

pub type BorderLineStyle = Keyword | datatypes.LineStyle
