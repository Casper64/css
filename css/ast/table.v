module ast

pub struct RuleNode {
pub:
	selectors []Node
	// specificity  Specificity
	declarations []Node
}

@[heap; minify]
pub struct Table {
pub mut:
	// front = least specific, back = most specific
	rules []RuleNode
	// at_rules []AtRuleNode
}

pub fn (mut t Table) insert_declarations(selector Selector, declarations []Node) {
	// specificity := Specificity.from_selector(selector)
	// node := RuleNode{
	// 	selectors: selector.children
	// 	specificity: specificity
	// 	declarations: declarations
	// }

	// t.rules << node
}

pub fn (mut t Table) sort_rules() {
	// t.rules.sort(|a, b| a.specificity <= b.specificity)
}
