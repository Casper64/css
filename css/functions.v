module css

import css.datatypes

pub enum UrlKind {
	link // http://
	data // data:image/png;base64,
	file // myFont.woff
	element // #IDofSVGpath
}

pub struct Url {
pub mut:
	kind  UrlKind
	value string
}

pub fn (u1 Url) == (u2 Url) bool {
	if u1.kind != u2.kind {
		return false
	}

	return u1.value == u2.value
}

pub struct Gradient {
pub mut:
	kind            datatypes.GradientKind
	directions      datatypes.GradientDirection
	gradient_values []GradientValue
}

pub struct GradientValue {
	color ColorValue
	size  ?DimensionValue
}
