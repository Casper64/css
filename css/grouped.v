module css

// set_grouped merges sub properties into its container property
// e.g. if `prop_name` = 'background-color':
// ```
// style_map['background'] gets set to css.Background{
//     color: value
// }
// ```
fn set_grouped(prop_name string, value Value, mut style_map map[string]Value) {
	if prop_name.starts_with('background-') {
		set_grouped_background(prop_name, value, mut style_map)
	} else if prop_name.starts_with('padding-') || prop_name.starts_with('margin-') {
		set_grouped_margin_padding(prop_name, value, mut style_map)
	} else if prop_name.starts_with('text-') {
		set_grouped_text(prop_name, value, mut style_map)
	} else if prop_name.starts_with('overflow-') {
		set_grouped_overflow(prop_name, value, mut style_map)
	} else if prop_name.starts_with('border-') {
		set_grouped_border(prop_name, value, mut style_map)
	} else {
		style_map[prop_name] = value
	}
}

fn set_grouped_background(prop_name string, value Value, mut style_map map[string]Value) {
	mut bg_val := (style_map['background'] or { Background{} }) as Background

	match prop_name {
		'background-color' {
			bg_val.color = value as ColorValue
		}
		'background-image' {
			bg_val.image = value as Image
		}
		else {}
	}

	style_map['background'] = bg_val
}

fn set_grouped_margin_padding(prop_name string, value Value, mut style_map map[string]Value) {
	// extract "padding" from "padding-left"
	short_name := prop_name.all_before('-')
	mut mgpd_val := (style_map[short_name] or { MarginPadding{} }) as MarginPadding

	match prop_name.all_after('-') {
		'top' {
			mgpd_val.top = value as DimensionValue
		}
		'right' {
			mgpd_val.right = value as DimensionValue
		}
		'bottom' {
			mgpd_val.right = value as DimensionValue
		}
		'left' {
			mgpd_val.left = value as DimensionValue
		}
		else {}
	}

	style_map[short_name] = mgpd_val
}

fn set_grouped_text(prop_name string, value Value, mut style_map map[string]Value) {
	mut txt_val := (style_map['text'] or { Text{} }) as Text

	match prop_name {
		'text-align' { txt_val.align = value as Keyword }
		'text-align-last' { txt_val.align_last = value as Keyword }
		'text-combine-upright' { txt_val.combine_upright = value as TextCombineUpright }
		'text-indent' { txt_val.indent = value as DimensionValue }
		'text-justify' { txt_val.justify = value as Keyword }
		'text-orientation' { txt_val.orientation = value as Keyword }
		'text-overflow' { txt_val.overflow = value as TextOverflow }
		'text-rendering' { txt_val.rendering = value as Keyword }
		'text-transform' { txt_val.transform = value as Keyword }
		'text-wrap' { txt_val.wrap = value as Keyword }
		else {}
	}

	style_map['text'] = txt_val
}

fn set_grouped_overflow(prop_name string, value Value, mut style_map map[string]Value) {
	mut overflow_val := (style_map['overflow'] or { Overflow{} }) as Overflow

	match prop_name {
		'overflow-x' { overflow_val.overflow_x = value as Keyword }
		'overflow-y' { overflow_val.overflow_y = value as Keyword }
		else {}
	}

	style_map['overflow'] = overflow_val
}

fn set_grouped_border(prop_name string, value Value, mut style_map map[string]Value) {
	mut border_val := (style_map['border'] or { Border{} }) as Border

	match prop_name {
		'border-color' {
			border_val.colors = value as BorderColors
		}
		'border-style' {
			border_val.styles = value as BorderStyles
		}
		'border-width' {
			border_val.width = value as DimensionValue
		}
		else {
			if prop_name.ends_with('-color') {
				border_val.colors = get_grouped_border_color(prop_name, value, style_map)
			}
		}
	}

	style_map['border'] = border_val
}

fn get_grouped_border_color(prop_name string, value Value, style_map map[string]Value) BorderColors {
	mut bcolor_val := BorderColors{}
	if v := style_map['border'] {
		bcolor_val = (v as Border).colors
	}

	match prop_name {
		'border-top-color' { bcolor_val.top = value as ColorValue }
		'border-right-color' { bcolor_val.right = value as ColorValue }
		'border-bottom-color' { bcolor_val.bottom = value as ColorValue }
		'border-left-color' { bcolor_val.left = value as ColorValue }
		else {}
	}

	return bcolor_val
}
