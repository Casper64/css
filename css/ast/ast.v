module ast

// implementing ast from https://github.com/csstree/csstree/blob/master/docs/ast.md
// following the mdn specs for css https://developer.mozilla.org/en-US/docs/Web/CSS
import css.token

// empty struct
pub struct Empty {}

// a css file
[heap]
pub struct StyleSheet {
pub:
	file_path string
pub mut:
	imports []ImportRule
	rules   []RuleType
}

pub struct Rule {
pub:
	pos token.Pos
pub mut:
	prelude Prelude
	block   Block
}

pub struct AtRule {
pub:
	pos token.Pos
pub mut:
	typ      AtType
	prelude  []Node
	children []Node
}

pub struct ImportRule {
pub:
	pos token.Pos
pub mut:
	typ   ImportType
	path  string
	layer ?string
	// supports []SupportRule
	media_query_list ?MediaQueryList
}

pub struct KeyframesRule {
pub:
	pos token.Pos
pub mut:
	percentage   string
	declarations []Node
}

pub struct Block {
pub:
	pos token.Pos
pub mut:
	declarations []Node
}

pub struct Declaration {
pub:
	pos token.Pos
pub mut:
	property  string
	important bool
	value     Value
}

pub struct Function {
pub:
	pos token.Pos
pub mut:
	name     string
	children []Node
}

pub struct Dimension {
pub:
	pos token.Pos
pub mut:
	value string
	unit  string
}

pub struct Hash {
pub:
	pos token.Pos
pub mut:
	value string
}

pub struct Ident {
pub:
	pos token.Pos
pub mut:
	name string
}

pub struct Number {
pub:
	pos token.Pos
pub mut:
	value string
}

pub struct Operator {
pub:
	pos token.Pos
pub mut:
	kind OperatorKind
}

pub fn (op Operator) == (other Operator) bool {
	return op.kind == other.kind
}

pub struct Parentheses {
pub:
	pos token.Pos
pub mut:
	children []Node
}

pub struct MediaFeature {
pub:
	pos token.Pos
pub mut:
	name  string
	value ?Node
}

pub struct MediaQuery {
pub:
	pos token.Pos
pub mut:
	children []Node
}

pub struct MediaQueryList {
pub:
	pos token.Pos
pub mut:
	children []MediaQuery
}

// Raw is used for some functions like `url()` or anything that can't be properly parsed
pub struct Raw {
pub:
	pos token.Pos
pub mut:
	value string
}

pub struct Value {
pub:
	pos token.Pos
pub mut:
	children []Node
}

pub struct SelectorList {
pub:
	pos token.Pos
pub mut:
	children []Node
}

pub struct Selector {
pub:
	pos token.Pos
pub mut:
	children []Node
}

// matches tests if the selectors in `s1` match the selectors in `nodes`:
// it checks if s1 is a subset of nodes
// the selectors children will be split on combinators and onl the last part is matched:
// .test > .other#id will only try to match for .other#id
pub fn (selector Selector) matches(other_selectors []Node) bool {
	mut last_nodes := []Node{}
	for n in selector.children {
		if n is Combinator {
			last_nodes.clear()
		} else {
			last_nodes << n
		}
	}

	for sub_s1 in last_nodes {
		if sub_s1 !in other_selectors {
			return false
		}
		// match sub_s1 {
		// 	IdSelector, ClassSelector, TypeSelector {
		// 		if sub_s1 !in s2.children {
		// 			return false
		// 		}
		// 	}
		// 	else {
		// 		return false
		// 	}
		// }
	}
	return true
}

pub struct ClassSelector {
pub:
	pos token.Pos
pub mut:
	name string
}

pub struct TypeSelector {
pub:
	pos token.Pos
pub mut:
	name string
}

pub struct IdSelector {
pub:
	pos token.Pos
pub mut:
	name string
}

pub struct Combinator {
pub:
	pos token.Pos
pub mut:
	kind string
}

pub struct PseudoClassSelector {
pub:
	pos token.Pos
pub mut:
	name     string
	children []Node
}

pub struct PseudoElementSelector {
pub:
	pos token.Pos
pub mut:
	name     string
	children []Node
}

pub struct AttributeSelector {
pub:
	pos token.Pos
pub mut:
	name    Ident
	matcher AttributeMatchType
	value   ?String
}

pub struct String {
pub:
	pos token.Pos
pub mut:
	value string
}

pub type Node = AtRule
	| AttributeSelector
	| Block
	| ClassSelector
	| Combinator
	| Declaration
	| Dimension
	| Empty
	| Function
	| Hash
	| IdSelector
	| Ident
	| ImportRule
	| KeyframesRule
	| MediaFeature
	| MediaQuery
	| MediaQueryList
	| NodeError
	| Number
	| Operator
	| Parentheses
	| Prelude
	| PseudoClassSelector
	| PseudoElementSelector
	| PseudoSelector
	| Raw
	| Rule
	| RuleType
	| Selector
	| SelectorList
	| String
	| StyleSheet
	| TypeSelector
	| Value

pub fn (node Node) children() []Node {
	return match node {
		AtRule {
			mut children := []Node{}
			children << node.prelude
			children << node.children
			children
		}
		Rule {
			mut children := []Node{}
			children << Node(node.prelude)
			children << node.block
			children
		}
		ImportRule {
			if l := node.media_query_list {
				[Node(l)]
			} else {
				[]Node{}
			}
		}
		Block, KeyframesRule {
			node.declarations
		}
		Declaration {
			[Node(node.value)]
		}
		Function, Parentheses, MediaQuery, Value, SelectorList, Selector, PseudoClassSelector,
		PseudoElementSelector {
			node.children
		}
		MediaFeature {
			if f := node.value {
				return [f]
			} else {
				[]Node{}
			}
		}
		MediaQueryList {
			mut children := []Node{}
			children << node.children
			children
		}
		else {
			[]Node{}
		}
	}
}

pub fn (node Node) pos() token.Pos {
	return match node {
		StyleSheet, Empty {
			token.Pos{}
		}
		Rule, AtRule, ImportRule, KeyframesRule, Block, Declaration, Function, Dimension, Hash,
		Ident, Number, Operator, Parentheses, MediaFeature, MediaQuery, MediaQueryList, Raw, Value,
		SelectorList, Selector, ClassSelector, TypeSelector, IdSelector, Combinator,
		PseudoClassSelector, PseudoElementSelector, AttributeSelector, String {
			node.pos
		}
		else {
			token.Pos{}
		}
	}
}
