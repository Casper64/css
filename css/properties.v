module css

import css.datatypes

// pub type Width = DimensionValue
// pub type Height = DimensionValue
// pub type Top = DimensionValue
// pub type Right = DimensionValue
// pub type Bottom = DimensionValue
// pub type Left = DimensionValue

// collection of properties listed here:
// https://developer.mozilla.org/en-US/docs/Web/CSS/background
pub struct Background {
pub mut:
	color ColorValue
}

// collection of properties listed here:
// https://developer.mozilla.org/en-US/docs/Web/CSS/border
pub struct Border {
pub mut:
	color ColorValue
	style datatypes.LineStyle
	width DimensionValue
}

// collection of properties listed here:
// https://developer.mozilla.org/en-US/docs/Web/CSS/margin
pub struct Margin {
pub mut:
	top    DimensionValue
	right  DimensionValue
	bottom DimensionValue
	left   DimensionValue
}

// collection of properties listed here:
// https://developer.mozilla.org/en-US/docs/Web/CSS/padding
pub struct Padding {
pub mut:
	top    DimensionValue
	right  DimensionValue
	bottom DimensionValue
	left   DimensionValue
}
