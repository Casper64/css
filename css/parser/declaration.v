module parser

import css.ast

pub fn (mut p Parser) parse_declaration_block() []ast.Node {
	mut declarations := []ast.Node{}

	for {
		if p.tok.kind == .rcbr || p.tok.kind == .eof {
			break
		}

		if decl := p.parse_declaration() {
			declarations << decl
		} else {
			// continue to next line
			for p.tok.kind !in [.semicolon, .rcbr, .eof] {
				p.next()
			}
			// prevent parser from getting stuck e.g. "color: green;;"
			if p.tok.kind == .semicolon {
				p.next()
			}

			if err is ast.NodeError {
				declarations << ast.Raw{
					value: p.lexer.get_lit(err.pos.pos, p.tok.pos().pos)
					pos: err.pos.extend(p.tok.pos())
				}
			} else {
				p.error(err.msg())
			}
		}
	}

	return declarations
}

pub fn (mut p Parser) parse_declaration() !ast.Declaration {
	defer {
		p.is_important = false
	}
	start_pos := p.tok.pos()

	if p.tok.kind != .name {
		// .test { ; }
		return p.error('unexpected token, expecting a property name or "}"')
	}

	property := p.tok.lit
	p.next()
	if p.tok.kind != .colon {
		// .test { color green; }
		return p.error('expecting ":" after a property name')
	}
	p.next()

	value_pos := p.tok.pos()
	values := p.parse_value_children()!
	mut value := ast.Value{
		pos: value_pos.extend(p.tok.pos())
		children: values
	}

	return ast.Declaration{
		pos: start_pos.extend(p.tok.pos())
		property: property
		value: value
		important: p.is_important
	}
}

pub fn (mut p Parser) parse_value_children() ![]ast.Node {
	mut children := []ast.Node{}
	for {
		match p.tok.kind {
			.exclamation {
				p.next()
				important_pos := p.prev_tok.pos().extend(p.tok.pos())
				// color: !asdsd;
				if p.tok.kind != .name || p.tok.lit != 'important' {
					return p.error_with_pos('expecting "!important"', important_pos)
				} else {
					// color: !important;
					if children.len == 0 {
						return p.error_with_pos('expecting a value before "!important"',
							important_pos)
					}
					p.is_important = true
					// color: red !important green;
					if p.peek_tok.kind != .semicolon && p.peek_tok.kind != .rcbr {
						return p.error_with_pos('expecting ";": "!important" has to be the last identifier in a declaration',
							important_pos)
					}
				}
			}
			.number {
				// if the tokens are directly next to each other it's a dimension: 100px
				// otherwise it's a number and something else: rgb(0 0 0)
				if p.tok.pos().pos + p.tok.pos().len == p.peek_tok.pos().pos
					&& (p.peek_tok.kind == .name || p.peek_tok.kind == .percentage) {
					children << p.parse_dimension()
				} else {
					children << ast.Number{
						pos: p.tok.pos()
						value: p.tok.lit
					}
				}
			}
			.name {
				if p.peek_tok.kind == .lpar {
					children << p.parse_function()!
				} else {
					children << ast.Ident{
						pos: p.tok.pos()
						name: p.tok.lit
					}
				}
			}
			.string {
				// e.g. content: "stuff";
				children << ast.String{
					pos: p.tok.pos()
					value: p.tok.lit
				}
			}
			.hash {
				hash_start := p.tok.pos()
				if p.peek_tok.kind != .color {
					// color: #;
					return p.error_with_pos('expecting a hex color', hash_start.extend(p.peek_tok.pos()))
				}
				p.next()
				p.validate_hex_color(p.tok.lit)!
				children << ast.Hash{
					pos: hash_start.extend(p.tok.pos())
					value: p.tok.lit
				}
			}
			// operators
			.comma {
				children << ast.Operator{
					pos: p.tok.pos()
					kind: .comma
				}
			}
			.plus {
				children << ast.Operator{
					pos: p.tok.pos()
					kind: .plus
				}
			}
			.minus {
				children << ast.Operator{
					pos: p.tok.pos()
					kind: .min
				}
			}
			.mul {
				children << ast.Operator{
					pos: p.tok.pos()
					kind: .mul
				}
			}
			.div {
				children << ast.Operator{
					pos: p.tok.pos()
					kind: .div
				}
			}
			.rcbr {
				// it's actually valid css syntax if you don't place a semicolon after the last
				// declaration in a block
				if p.prefs.is_strict {
					return p.error('expecting ";"')
				}
				return children
			}
			.lpar {
				children << p.parse_parentheses()!
			}
			.rpar {
				if p.parentheses_depth <= 0 {
					return p.error('unexpected ")" there is no matching opening parenthesis')
				}
				return children
			}
			.semicolon {
				if p.parentheses_depth > 0 {
					// :not(.test  { color: green; }
					//           ^
					return p.error('unexpected ";" inside a function, expecting ")". Did you forget to close the parentheses?')
				}
				break
			}
			else {
				return p.unexpected()
			}
		}
		p.next()
	}
	// skip ';'
	p.next()

	return children
}

pub fn (mut p Parser) parse_dimension() ast.Dimension {
	start_pos := p.tok.pos()
	value := p.tok.lit
	p.next()
	unit := p.tok.lit
	return ast.Dimension{
		pos: start_pos.extend(p.tok.pos())
		value: value
		unit: unit
	}
}

pub fn (mut p Parser) parse_parentheses() !ast.Parentheses {
	par_start := p.tok.pos()
	p.next()

	p.parentheses_depth++
	children := p.parse_value_children()!
	p.parentheses_depth--

	return ast.Parentheses{
		pos: par_start.extend(p.tok.pos())
		children: children
	}
}

pub fn (mut p Parser) parse_function() !ast.Function {
	fn_start := p.tok.pos()

	fn_name := p.tok.lit
	// skip "("
	p.next()
	p.next()
	p.parentheses_depth++
	defer {
		p.parentheses_depth--
	}
	children := p.parse_value_children()!

	if p.tok.kind != .rpar {
		return p.error('expecting closing parenthesis ")" not ${p.tok.kind}')
	}
	return ast.Function{
		pos: fn_start.extend(p.tok.pos())
		name: fn_name
		children: children
	}
}

// returns error if not valid
pub fn (mut p Parser) validate_hex_color(val string) ! {
	if val.to_lower().contains_only('abcdef0123456789') == false {
		return p.error('invalid hex color')
	}

	// TODO: handle this in checker?

	// check if the next token is next to the color and if that token is not ; or , or ! or }
	// that means it is an invalid hex color
	// #0000zz
	// #0000width
	if p.peek_tok.kind !in [.comma, .semicolon, .rcbr, .exclamation]
		&& p.peek_tok.pos == p.tok.pos + p.tok.len {
		return p.error_with_pos('invalid hex color', p.tok.pos().extend(p.peek_tok.pos()))
	}
}
