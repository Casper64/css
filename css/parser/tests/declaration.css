
.test {
    ;
    color green;
    color: rgba 0 0 0);
}

.test {
    color: red;
}
