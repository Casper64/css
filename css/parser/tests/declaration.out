css/parser/tests/declaration.css:3:5: error: unexpected token, expecting a property name or "}"
    1 | 
    2 | .test {
    3 |     ;
      |     ^
    4 |     color green;
    5 |     color: rgba 0 0 0);
css/parser/tests/declaration.css:4:11: error: expecting ":" after a property name
    2 | .test {
    3 |     ;
    4 |     color green;
      |           ~~~~~
    5 |     color: rgba 0 0 0);
    6 | }
css/parser/tests/declaration.css:5:22: error: unexpected ")" there is no matching opening parenthesis
    3 |     ;
    4 |     color green;
    5 |     color: rgba 0 0 0);
      |                      ^
    6 | }
    7 |